research/delf/delf/python/training/global_features/train.py

```diff
@@ -0,0 +1,719 @@
+
+    for epoch in range(start_epoch, FLAGS.epochs + 1):
+      # Set manual seeds per epoch.
+      np.random.seed(epoch)
+      tf.random.set_seed(epoch)
+
+      # Find hard-negatives.
+      # While hard-positive examples are fixed during the whole training
+      # process and are randomly chosen from every epoch; hard-negatives
+      # depend on the current CNN parameters and are re-mined once per epoch.
+      avg_neg_distance = train_dataset.create_epoch_tuples(model,
+                                                           model_directory)
+
+      def train_gen():
+        return (inst for inst in train_dataset)
+
+      train_loader = tf.data.Dataset.from_generator(train_gen,
+                                                    output_types=tuple(
+                                                            output_types))
+
+      loss = train_val(loader=iter(train_loader), model=model,
+                       criterion=criterion, optimizer=optimizer,
+                       epoch=epoch)
+
+      # Write a scalar summary.
+      tf.summary.scalar('train_epoch_loss', loss, step=epoch)
+      # Forces summary writer to send any buffered data to storage.
+      writer.flush()
+
+      # Evaluate on validation set.
+      if FLAGS.val and (epoch % FLAGS.test_freq == 0 or epoch == 1):
+        avg_neg_distance = val_dataset.create_epoch_tuples(model,
+                                                           model_directory)
+
+        def val_gen():
+          return (inst for inst in val_dataset)
+
+        val_loader = tf.data.Dataset.from_generator(val_gen,
+                                                    output_types=tuple(
+                                                            output_types))
+
+        loss = train_val(loader=iter(val_loader), model=model,
+                         criterion=criterion, optimizer=None,
+                         epoch=epoch, train=False)
+        tf.summary.scalar('val_epoch_loss', loss, step=epoch)
+        writer.flush()
+
+      # Evaluate on test datasets every test_freq epochs.
+      if epoch == 1 or epoch % FLAGS.test_freq == 0:
+        test(FLAGS.test_datasets, model, writer=writer, epoch=epoch,
+             model_directory=model_directory)
+
+      # Saving checkpoints and model weights.
+      try:
+        save_path = manager.save(checkpoint_number=epoch)
+        global_features_utils.debug_and_log(
+                'Saved ({}) at {}'.format(epoch, save_path))
+
+        filename = os.path.join(model_directory,
+                                'checkpoint_epoch_{}.h5'.format(epoch))
+        model.save_weights(filename, save_format='h5')
+        global_features_utils.debug_and_log(
+                'Saved weights ({}) at {}'.format(epoch, filename))
+      except Exception as ex:
+        global_features_utils.debug_and_log(
+                'Could not save checkpoint: {}'.format(ex))
+
+
+def grad(criterion, model, input, target):
```

Comment:
for this and the following functions: let's maybe move them into a separate file? This script is already really long. It could be something like "train_utils.py". I will take a look at the functions once you move them.